components/generators/kubernetes/__init__.py

```diff
@@ -0,0 +1,667 @@
+        # TODO(ademaria) This name mangling is here just to simplify diff.
+        # Change it once done
+        component_name = self.kwargs.name
+        self.kwargs.name = "{}-metrics".format(component_name)
+
+        super().body()
+        name = self.kwargs.name
+        component = self.kwargs.component
+        self.add_namespace(inv.parameters.namespace)
+        self.root.spec.endpoints = component.service_monitors.endpoints
+        self.root.spec.jobLabel = name
+        self.root.spec.namespaceSelector.matchNames = [component_name]
+        self.root.spec.selector.matchLabels["name"] = component_name
+
+
+class MutatingWebhookConfiguration(k8s.Base):
+    def new(self):
+        self.kwargs.apiVersion = "admissionregistration.k8s.io/v1beta1"
+        self.kwargs.kind = "MutatingWebhookConfiguration"
+        super().new()
+        self.need("component")
+
+    def body(self):
+        super().body()
+        name = self.kwargs.name
+        component = self.kwargs.component
+        self.root.webhooks = component.webhooks
+
+
+class ClusterRole(k8s.Base):
+    def new(self):
+        self.kwargs.apiVersion = "rbac.authorization.k8s.io/v1beta1"
+        self.kwargs.kind = "ClusterRole"
+        super().new()
+        self.need("component")
+
+    def body(self):
+        super().body()
+        name = self.kwargs.name
+        component = self.kwargs.component
+        self.root.rules = component.cluster_role.rules
+
+
+class ClusterRoleBinding(k8s.Base):
+    def new(self):
+        self.kwargs.apiVersion = "rbac.authorization.k8s.io/v1beta1"
+        self.kwargs.kind = "ClusterRoleBinding"
+        super().new()
+        self.need("component")
+
+    def body(self):
+        super().body()
+        name = self.kwargs.name
+        component = self.kwargs.component
+        self.root.roleRef = component.cluster_role.binding.roleRef
+        self.root.subjects = component.cluster_role.binding.subjects
+
+
+def get_components():
+    if 'components' in inv.parameters:
+        generator_defaults = inv.parameters.generators.manifest.default_config
+
+        for name, component in inv.parameters.components.items():
+            if "application" in component:
+                application_defaults = inv.parameters.applications.get(
+                    component.application, {}).get(
+                    'component_defaults', {})
+                merge(generator_defaults, application_defaults)
+                if component.get("type", "undefined") in component.globals:
+                    merge(application_defaults, component.globals.get(component.type, {}))
+                merge(application_defaults, component)
+
+            merge(generator_defaults, component)
+            component.name = name
+            yield name, component
+
+
+def generate_docs():
+    obj = BaseObj()
+    for name, component in get_components():
+        obj.root["{}-readme.md".format(name)] = j2('templates/docs/service_component.md.j2',
```

Comment:
`compile` fails if this file isn't present, i don't think this should be hardcoded

Reply:
addressed.